scrapy/utils/gz.py

```diff
@@ -4,30 +4,38 @@
     from cStringIO import StringIO as BytesIO
 except ImportError:
     from io import BytesIO
-
+from io import UnsupportedOperation
 from gzip import GzipFile
 
+class ReadOneGzipFile(GzipFile):
+    def readone(self, size=-1):
+        try:
+            return self.read1(size)
+        except UnsupportedOperation:
+            return self.read(size)
 
 def gunzip(data):
     """Gunzip the given data and return as much data as possible.
 
     This is resilient to CRC checksum errors.
     """
-    f = GzipFile(fileobj=BytesIO(data))
+    f = ReadOneGzipFile(fileobj=BytesIO(data))
     output = b''
     chunk = b'.'
     while chunk:
         try:
-            chunk = f.read(8196)
+            chunk = f.readone(8196)
             output += chunk
         except (IOError, EOFError, struct.error):
             # complete only if there is some data, otherwise re-raise
             # see issue 87 about catching struct.error
             # some pages are quite small so output is '' and f.extrabuf
             # contains the whole page content
-            if output or f.extrabuf:
-                output += f.extrabuf
-                break
+            if output or getattr(f, 'extrabuf', None):
+                try:
+                    output += f.extrabuf
+                finally:
+                    break
```

Comment:
@kmike ,
with Python 3.5, with current code, the 3 failure tests' `read()` call correctly raises an exception, but you also get an exception for `extrabuf` in all 3 cases (`AttributeError: 'GzipFile' object has no attribute 'extrabuf'`):

`GunzipTest.test_gunzip_no_gzip_file_raises --> OSError("Not a gzipped file (b'<?')",)` and  output is `b''`

`GunzipTest.test_gunzip_truncated --> EOFError('Compressed file ended before the end-of-stream marker was reached',),` but output contains some data but not all what the tests expects

`GunzipTest.test_gunzip_truncated_short --> EOFError('Compressed file ended before the end-of-stream marker was reached',)` and output is empty `b''`

after adding protection against missing `extrabuf`, only `GunzipTest.test_gunzip_no_gzip_file_raises` passes additionally (with empty output) ; the other 2 fail for missing data in output